File: backend/services/query_engine.py
import re
import json
from typing import List, Dict, Any, Optional
from functools import cmp_to_key
from sqlmodel import Session, select
from database import DatabaseRecord, Database

def evaluate_formula(expression: str, values: Dict[str, Any]) -> Any:
    """Safely evaluates a basic mathematical/string formula expression in Python."""
    if not expression:
        return 0
        
    # Replace prop("...") or prop('...') with its value
    def replace_prop(match):
        prop_name = match.group(1)
        val = values.get(prop_name, 0)
        if val is None:
            return "0"
        if isinstance(val, (int, float)):
            return str(val)
        return repr(str(val))
    
    # Match prop("name") or prop('name')
    expr = re.sub(r"prop\(['\"](.*?)['\"]\)", replace_prop, expression)
    
    # Sanitize: allow only standard arithmetic characters, parentheses, numbers, and strings
    # Ensure no dangerous commands are run
    cleaned = re.sub(r"[0-9\+\-\*\/\.\(\)\s\'\"]", "", expr)
    if cleaned == "":
        try:
            return eval(expr, {"__builtins__": None}, {})
        except Exception:
            return 0
    return expr

def evaluate_rollups(record_values: Dict[str, Any], properties: List[Dict[str, Any]], session: Session) -> Dict[str, Any]:
    """Computes rollups and formulas for a record by traversing relations."""
    computed = dict(record_values)
    
    for prop in properties:
        prop_name = prop["name"]
        prop_type = prop["type"]
        
        # 1. Rollup property calculations
        if prop_type == "rollup":
            relation_field = prop.get("relation_property")
            target_field = prop.get("target_property")
            agg_func = prop.get("aggregation", "show_original")
            
            # Get related record IDs
            related_ids = record_values.get(relation_field, [])
            if not isinstance(related_ids, list):
                related_ids = [related_ids] if related_ids else []
                
            if not related_ids:
                computed[prop_name] = 0 if agg_func in ["sum", "average", "count"] else ""
                continue
                
            # Query related records
            related_records = session.exec(
                select(DatabaseRecord).where(DatabaseRecord.id.in_(related_ids))
            ).all()
            
            # Extract target values
            vals = []
            for r in related_records:
                vals_dict = json.loads(r.property_values)
                v = vals_dict.get(target_field)
                if v is not None:
                    vals.append(v)
            
            # Aggregate
            if agg_func == "sum":
                computed[prop_name] = sum(float(x) for x in vals if str(x).replace(".", "", 1).isdigit())
            elif agg_func == "average":
                num_vals = [float(x) for x in vals if str(x).replace(".", "", 1).isdigit()]
                computed[prop_name] = sum(num_vals) / len(num_vals) if num_vals else 0
            elif agg_func == "count":
                computed[prop_name] = len(vals)
            else: # show_original
                computed[prop_name] = vals[0] if len(vals) == 1 else vals
                
        # 2. Formula property calculations
        elif prop_type == "formula":
            formula_expr = prop.get("formula_expression", "")
            computed[prop_name] = evaluate_formula(formula_expr, computed)
            
    return computed

def matches_filter(record: Dict[str, Any], filter_group: Dict[str, Any]) -> bool:
    """Recursively checks if a record matches nested filter groups."""
    if not filter_group or "operator" not in filter_group:
        return True
        
    operator = filter_group.get("operator", "and").lower()
    rules = filter_group.get("rules", [])
    
    if not rules:
        return True
        
    results = []
    for rule in rules:
        if "operator" in rule:
            # Nested filter group
            results.append(matches_filter(record, rule))
        else:
            # Rule comparison
            prop_name = rule.get("property")
            cond = rule.get("condition", "equals")
            target_val = rule.get("value")
            
            val = record.get(prop_name)
            
            if cond == "equals":
                results.append(str(val).lower() == str(target_val).lower())
            elif cond == "contains":
                results.append(str(target_val).lower() in str(val).lower())
            elif cond == "starts_with":
                results.append(str(val).lower().startswith(str(target_val).lower()))
            elif cond == "is_empty":
                results.append(val is None or val == "" or val == [])
            elif cond == "is_not_empty":
                results.append(not (val is None or val == "" or val == []))
            elif cond == "greater_than":
                try:
                    results.append(float(val) > float(target_val))
                except Exception:
                    results.append(False)
            elif cond == "less_than":
                try:
                    results.append(float(val) < float(target_val))
                except Exception:
                    results.append(False)
            else:
                results.append(True)
                
    if operator == "or":
        return any(results)
    return all(results)

def sort_records(records: List[Dict[str, Any]], sort_rules: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
    """Sorts record dicts according to priority list of sort rules."""
    if not sort_rules:
        return records
        
    def compare(r1, r2):
        for rule in sort_rules:
            prop_name = rule.get("property")
            direction = rule.get("direction", "asc").lower()
            
            v1 = r1.get(prop_name)
            v2 = r2.get(prop_name)
            
            if v1 is None: v1 = ""
            if v2 is None: v2 = ""
            
            if v1 != v2:
                if direction == "desc":
                    return 1 if v1 < v2 else -1
                else:
                    return -1 if v1 < v2 else 1
        return 0
        
    return sorted(records, key=cmp_to_key(compare))

def group_records(records: List[Dict[str, Any]], grouping_rules: Dict[str, Any]) -> Dict[str, Any]:
    """Arranges flat records into structured subgroups based on primary/sub keys."""
    group_by = grouping_rules.get("group_by")
    subgroup_by = grouping_rules.get("subgroup_by")
    
    if not group_by:
        return {"grouped": False, "records": records}
        
    grouped = {}
    for r in records:
        g_val = r.get(group_by)
        if isinstance(g_val, list):
            g_val = g_val[0] if g_val else "None"
        g_val = str(g_val) if g_val is not None else "None"
        
        if g_val not in grouped:
            grouped[g_val] = []
        grouped[g_val].append(r)
        
    if subgroup_by:
        for g_val, g_recs in grouped.items():
            subgrouped = {}
            for r in g_recs:
                sg_val = r.get(subgroup_by)
                if isinstance(sg_val, list):
                    sg_val = sg_val[0] if sg_val else "None"
                sg_val = str(sg_val) if sg_val is not None else "None"
                
                if sg_val not in subgrouped:
                    subgrouped[sg_val] = []
                subgrouped[sg_val].append(r)
            grouped[g_val] = subgrouped
            
    return {
        "grouped": True,
        "group_by": group_by,
        "subgroup_by": subgroup_by,
        "groups": grouped
    }

class QueryEngine:
    @staticmethod
    def execute_view(session: Session, view_id: str) -> Dict[str, Any]:
        """Query primary database, calculate properties, apply sorting, filtering and grouping."""
        from database import DatabaseView
        
        # 1. Fetch View Config
        view = session.query(DatabaseView).filter(DatabaseView.id == view_id).first()
        if not view:
            raise ValueError(f"View {view_id} not found.")
            
        # 2. Fetch Database properties
        db = session.query(Database).filter(Database.id == view.database_id).first()
        properties = json.loads(db.properties) if db else []
        
        # 3. Load Database Records
        raw_records = session.exec(
            select(DatabaseRecord).where(DatabaseRecord.database_id == view.database_id)
        ).all()
        
        # 4. Parse property values and evaluate formulas/rollups
        computed_records = []
        for r in raw_records:
            vals = json.loads(r.property_values)
            # Add implicit ID field
            vals["id"] = r.id
            # Resolve computed fields
            computed_vals = evaluate_rollups(vals, properties, session)
            computed_records.append(computed_vals)
            
        # 5. Apply Filter AST
        filters = json.loads(view.filters) if view.filters else {}
        filtered_records = [r for r in computed_records if matches_filter(r, filters)]
        
        # 6. Apply Sort priority
        sorts = json.loads(view.sorts) if view.sorts else []
        sorted_records = sort_records(filtered_records, sorts)
        
        # 7. Apply Grouping (Columns & Swimlanes)
        grouping = json.loads(view.grouping) if view.grouping else {}
        grouped_result = group_records(sorted_records, grouping)
        
        # Append meta
        grouped_result["view_name"] = view.name
        grouped_result["layout_type"] = view.layout_type
        grouped_result["visible_properties"] = json.loads(view.visible_properties) if view.visible_properties else []
        
        return grouped_result

File: backend-js/src/views/queryEngine/QueryEngine.ts
import { evaluateFormula } from "./formulaEvaluator.js";
import type {
  DatabaseProperty,
  DatabaseSchema,
  DatabaseViewConfig,
  ExecuteViewResult,
  FilterGroup,
  FilterRule,
  GroupingRules,
  RawDatabaseRecord,
  SortRule,
} from "./types.js";

function parsePropertyValues(record: RawDatabaseRecord): Record<string, unknown> {
  if (typeof record.property_values === "string") {
    return JSON.parse(record.property_values) as Record<string, unknown>;
  }
  return { ...record.property_values };
}

function isFilterGroup(rule: FilterRule | FilterGroup): rule is FilterGroup {
  return "operator" in rule && Array.isArray((rule as FilterGroup).rules);
}

/**
 * Recursively checks if a record matches nested filter groups.
 */
export function matchesFilter(
  record: Record<string, unknown>,
  filterGroup: FilterGroup,
): boolean {
  if (!filterGroup || !("operator" in filterGroup)) {
    return true;
  }

  const operator = (filterGroup.operator ?? "and").toLowerCase();
  const rules = filterGroup.rules ?? [];

  if (rules.length === 0) {
    return true;
  }

  const results: boolean[] = [];

  for (const rule of rules) {
    if (isFilterGroup(rule)) {
      results.push(matchesFilter(record, rule));
      continue;
    }

    const propName = rule.property;
    const cond = rule.condition ?? "equals";
    const targetVal = rule.value;
    const val = record[propName];

    switch (cond) {
      case "equals":
        results.push(String(val).toLowerCase() === String(targetVal).toLowerCase());
        break;
      case "contains":
        results.push(String(val).toLowerCase().includes(String(targetVal).toLowerCase()));
        break;
      case "starts_with":
        results.push(String(val).toLowerCase().startsWith(String(targetVal).toLowerCase()));
        break;
      case "is_empty":
        results.push(val === null || val === undefined || val === "" || (Array.isArray(val) && val.length === 0));
        break;
      case "is_not_empty":
        results.push(!(val === null || val === undefined || val === "" || (Array.isArray(val) && val.length === 0)));
        break;
      case "greater_than":
        try {
          results.push(Number(val) > Number(targetVal));
        } catch {
          results.push(false);
        }
        break;
      case "less_than":
        try {
          results.push(Number(val) < Number(targetVal));
        } catch {
          results.push(false);
        }
        break;
      default:
        results.push(true);
    }
  }

  if (operator === "or") {
    return results.some(Boolean);
  }
  return results.every(Boolean);
}

/**
 * Sorts record dicts according to priority list of sort rules.
 */
export function sortRecords(
  records: Record<string, unknown>[],
  sortRules: SortRule[],
): Record<string, unknown>[] {
  if (!sortRules || sortRules.length === 0) {
    return records;
  }

  return [...records].sort((r1, r2) => {
    for (const rule of sortRules) {
      const propName = rule.property;
      const direction = (rule.direction ?? "asc").toLowerCase();

      let v1: unknown = r1[propName];
      let v2: unknown = r2[propName];

      if (v1 === null || v1 === undefined) v1 = "";
      if (v2 === null || v2 === undefined) v2 = "";

      if (v1 !== v2) {
        const cmp = String(v1).localeCompare(String(v2));
        if (direction === "desc") {
          return -cmp;
        }
        return cmp;
      }
    }
    return 0;
  });
}

/**
 * Arranges flat records into structured subgroups based on primary/sub keys.
 */
export function groupRecords(
  records: Record<string, unknown>[],
  groupingRules: GroupingRules,
): Pick<ExecuteViewResult, "grouped" | "group_by" | "subgroup_by" | "groups" | "records"> {
  const groupBy = groupingRules.group_by;
  const subgroupBy = groupingRules.subgroup_by;

  if (!groupBy) {
    return { grouped: false, records };
  }

  const grouped: Record<string, unknown> = {};

  for (const record of records) {
    let gVal: unknown = record[groupBy];
    if (Array.isArray(gVal)) {
      gVal = gVal.length > 0 ? gVal[0] : "None";
    }
    const groupKey = gVal != null ? String(gVal) : "None";

    if (!grouped[groupKey]) {
      grouped[groupKey] = [];
    }
    (grouped[groupKey] as Record<string, unknown>[]).push(record);
  }

  if (subgroupBy) {
    for (const [gVal, gRecs] of Object.entries(grouped)) {
      const subgrouped: Record<string, unknown> = {};
      for (const record of gRecs as Record<string, unknown>[]) {
        let sgVal: unknown = record[subgroupBy];
        if (Array.isArray(sgVal)) {
          sgVal = sgVal.length > 0 ? sgVal[0] : "None";
        }
        const subKey = sgVal != null ? String(sgVal) : "None";

        if (!subgrouped[subKey]) {
          subgrouped[subKey] = [];
        }
        (subgrouped[subKey] as Record<string, unknown>[]).push(record);
      }
      grouped[gVal] = subgrouped;
    }
  }

  return {
    grouped: true,
    group_by: groupBy,
    subgroup_by: subgroupBy ?? null,
    groups: grouped,
    records: [],
  };
}

/**
 * Rollup stub — returns safe empty values without throwing.
 * Full rollup aggregation deferred to v2.
 */
export function evaluateRollupStub(
  recordValues: Record<string, unknown>,
  prop: DatabaseProperty,
): unknown {
  const aggFunc = prop.aggregation ?? "show_original";
  const relationField = prop.relation_property ?? "";
  const relatedIds = recordValues[relationField];

  const ids = Array.isArray(relatedIds)
    ? relatedIds
    : relatedIds
      ? [relatedIds]
      : [];

  if (ids.length === 0) {
    return aggFunc === "sum" || aggFunc === "average" || aggFunc === "count" ? 0 : "";
  }

  return aggFunc === "count" ? 0 : "";
}

/**
 * Computes formula and rollup properties for a record.
 */
export function evaluateComputedFields(
  recordValues: Record<string, unknown>,
  properties: DatabaseProperty[],
): Record<string, unknown> {
  const computed = { ...recordValues };

  for (const prop of properties) {
    const propName = prop.name;
    const propType = prop.type;

    if (propType === "rollup") {
      computed[propName] = evaluateRollupStub(recordValues, prop);
    } else if (propType === "formula") {
      const formulaExpr = prop.formula_expression ?? "";
      computed[propName] = evaluateFormula(formulaExpr, computed);
    }
  }

  return computed;
}

/**
 * Executes the query engine pipeline for a view configuration.
 *
 * Pure function — no HTTP or database I/O.
 */
export function executeView(
  viewConfig: DatabaseViewConfig,
  records: RawDatabaseRecord[],
  databases: DatabaseSchema[],
): ExecuteViewResult {
  const db = databases.find((d) => d.id === viewConfig.database_id);
  const properties = db?.properties ?? [];

  const computedRecords: Record<string, unknown>[] = [];

  for (const record of records) {
    const vals = parsePropertyValues(record);
    vals.id = record.id;
    computedRecords.push(evaluateComputedFields(vals, properties));
  }

  const filteredRecords = computedRecords.filter((r) => matchesFilter(r, viewConfig.filters));
  const sortedRecords = sortRecords(filteredRecords, viewConfig.sorts);
  const groupedResult = groupRecords(sortedRecords, viewConfig.grouping);

  return {
    ...groupedResult,
    records: groupedResult.grouped ? [] : groupedResult.records,
    view_name: viewConfig.name,
    layout_type: viewConfig.layout_type,
    visible_properties: viewConfig.visible_properties,
  };
}
